design/merge-refactor.md

```diff
@@ -0,0 +1,26 @@
+# Design: Refactor the Merge Operation
+
+## Problem Description
+Currently, as part of our design:
+* Apply is used for both commit and merge
+* Merge iterator is implemented using the Diff iterator
+
+Our code for the merge operation got to a level where it's too complex and it is very hard to follow when inserting changes.
```

Comment:
Think this is the point where we need to explain that apply/merge uses an iterator which communicates a stream of changes that the diff found.
Working with changes that we identify in a different level of the api leave small space to optimize in the space.